pipelines/batch/src/main/java/com/google/fhir/analytics/metrics/FlinkPipelineMetrics.java

```diff
@@ -0,0 +1,220 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.fhir.analytics.metrics;
+
+import com.google.fhir.analytics.MetricsConstants;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import org.apache.beam.runners.core.metrics.MetricsContainerStepMap;
+import org.apache.beam.runners.flink.metrics.FlinkMetricContainer;
+import org.apache.beam.sdk.metrics.MetricNameFilter;
+import org.apache.beam.sdk.metrics.MetricQueryResults;
+import org.apache.beam.sdk.metrics.MetricResult;
+import org.apache.beam.sdk.metrics.MetricResults;
+import org.apache.beam.sdk.metrics.MetricsFilter;
+import org.apache.flink.core.execution.JobClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of the {@link PipelineMetrics} class. This class is used to retrieve the metrics
+ * of the pipelines for the undergoing batch.
+ *
+ * <p>This class assumes that only one batch of pipelines run at any point in time and hence it
+ * maintains a static list of {@link JobClient} and {@link CumulativeMetrics}, which at any point
+ * signifies the list of jobs currently running and the cumulative metrics collected so far for the
+ * completed jobs. The user of this class has to clear these metrics when the batch ends
+ */
+public class FlinkPipelineMetrics implements PipelineMetrics {
+
+  private static final Logger logger = LoggerFactory.getLogger(FlinkPipelineMetrics.class);
+
+  private static ConcurrentHashMap<String, JobClient> jobClientMap = new ConcurrentHashMap<>();
+
+  private static CumulativeMetrics cumulativeMetrics = new CumulativeMetrics(0l, 0l, 0l);
+
+  /**
+   * This method returns {@link CumulativeMetrics} of the all the completed jobs and the currently
+   * running ones for the ongoing batch.
+   */
+  @Override
+  public synchronized CumulativeMetrics getCumulativeMetricsForOngoingBatch() {
+    List<MetricQueryResults> onGoingMetricQueryResults = getOngoingMetricQueryResults();
+    return getUpdatedCumulativeMetrics(cumulativeMetrics, onGoingMetricQueryResults);
+  }
+
+  /**
+   * Clears and resets all the {@link JobClient} and {@link CumulativeMetrics} collected so far for
+   * the batch.
+   */
+  @Override
+  public synchronized void clearAllMetrics() {
+    jobClientMap.clear();
+    cumulativeMetrics = new CumulativeMetrics(0l, 0l, 0l);
+  }
+
+  /**
+   * This method is used to initialise the {@link CumulativeMetrics} with the {@code
+   * totalNoOfResources}
+   *
+   * @param totalNoOfResources the total number of resources to be processed in the current batch
+   */
+  @Override
+  public synchronized void setTotalNoOfResources(long totalNoOfResources) {
```

Comment:
Probably this might give a better clarity that this tracks the total number of all resources in the batch rather the resources in one pipeline (debatable though).

Reply:
Sure; maybe only do the `No` -> `Num` suggestion then.

Reply:
Done